### PR TITLE
Add CORS handler that permits all POST requests

### PR DESCRIPTION
**Pull request by cstahl1**
Add CORS handler that will always permit POST requests from Origin *.

This can be expanded upon as more sophisticated logic is required and
is just a stand-in.

### CODE REVIEW

File: pkg/collector/cors.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// CORS intercepts CORS OPTIONS requests and sends a fixed set of headers
+// that allows all requests
+type CORS struct {
+	handler http.Handler
+}
+
+// ServeHTTP delegates non-OPTIONS requests to the contained handler and
+// gives a fixed response for OPTIONS requests that allows all incoming
+// CORS requests
+func (c *CORS) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (cstahl1) [reply pkg/collector/cors.go]: I would argue that the collector won't be compliant, not the pipeline. My understanding of the pipeline is that it takes in requests that have a report upload and then processes them in some manner. We could expand the scope of the pipeline, but what do we gain by that?

File: pkg/collector/cors.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	if r.Method == "OPTIONS" {
+		w.Header().Set("Access-Control-Allow-Method", "POST")
```
[the remark this answers is not in the document]
Comment (cstahl1) [reply pkg/collector/cors.go]: Done

File: pkg/collector/cors.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcreager) [reply pkg/collector/cors.go]: > what do we gain by that?

I think it's a usability win and a code simplicity win.  I can't envision a case where you'd want to register a `Pipeline` as an HTTP handler and *not* want the CORS logic, and so it seems better to optimize for that case (just register your pipeline as a handler and you're done).

> My understanding of the pipeline is that it takes in requests that have a report upload and then processes them in some manner.

That's definitely true of `ProcessReports`, and I'm happy with how we ended up factoring that into a separate method for exactly this reason.  But the existing `ServeHTTP` is what takes that method and turns it into a Reporting HTTP handler, which must include CORS behavior to be compliant.

[the remark this answers is not in the document]
Comment (cstahl1) [reply pkg/collector/cors.go]: Done.

File: pkg/collector/pipeline.go [hunk 3]
```diff
@@ -112,8 +112,19 @@ func (p *Pipeline) ProcessReports(w http.ResponseWriter, r *http.Request) *Repor
 	return &reports
 }
 
+// serveCORS handles OPTIONS requests by allowing POST requrests with a
```
Comment (dcreager) [inline pkg/collector/pipeline.go]: nit (spelling): requests

File: pkg/collector/pipeline_test.go [hunk 4]
```diff
@@ -29,7 +29,23 @@ import (
...
 var validNelReportPath = filepath.Clean("../pipelinetest/testdata/reports/valid-nel-report.json")
 
-func TestIgnoreNonPOST(t *testing.T) {
+func TestRespondsToOptionsRequest(t *testing.T) {
+	pipeline := collector.NewPipeline(pipelinetest.NewSimulatedClock())
+	request := httptest.NewRequest("OPTIONS", "https://example.com/upload", bytes.NewReader([]byte("")))
+	response := httptest.NewRecorder()
+	pipeline.ServeHTTP(response, request)
+	if want, got := "POST", response.Header().Get("Access-Control-Allow-Methods"); got != want {
+		t.Errorf("response.Header().Get(\"Access-Control-Allow-Method\"): got %v, want %v", got, want)
```
Comment (dcreager) [inline pkg/collector/pipeline_test.go]: One remaining singular → plural

File: pkg/collector/pipeline_test.go [hunk 5]
```diff
@@ -29,7 +29,23 @@ import (
 
 var validNelReportPath = filepath.Clean("../pipelinetest/testdata/reports/valid-nel-report.json")
 
-func TestIgnoreNonPOST(t *testing.T) {
+func TestRespondsToOptionsRequest(t *testing.T) {
+	pipeline := collector.NewPipeline(pipelinetest.NewSimulatedClock())
+	request := httptest.NewRequest("OPTIONS", "https://example.com/upload", bytes.NewReader([]byte("")))
+	response := httptest.NewRecorder()
+	pipeline.ServeHTTP(response, request)
+	if want, got := "POST", response.Header().Get("Access-Control-Allow-Methods"); got != want {
```
Comment (dcreager) [inline pkg/collector/pipeline_test.go]: Does it work to change this to:

``` go
if want := "POST"; got := /* etc */; got != want {
```

I'm half-convinced that that's more readable.  What do you think?

**Comment by dcreager** [comment]
Oh and make sure to put

```
Closes: #27
```

at the end of the commit message when you squash merge.

File: pkg/collector/pipeline_test.go (hunk 5 above)
Comment (cstahl1) [reply pkg/collector/pipeline_test.go]: I'm completely convinced; I had that originally and it didn't compile. :(

**Pull request merged**